File: typescript/bob/bob.ts

import {isNumber} from "util";

export default class Bob {
    hey(phrase: string): string {
        const value = phrase.trim()
        if (this.isSilence(value)) { return "Fine. Be that way!" }
        else if (this.isYell(value)) { return "Whoa, chill out!" }
        else if (this.isQuestion(value)) { return "Sure." }
        else { return "Whatever." }
    }

    isQuestion(value: string): boolean {
        return value.endsWith("?")
    }

    isYell(value: string): boolean {
        const characters = value.split('')
        const isUpperCase = characters.every((v) => v.toUpperCase() === v)
        const isNumbersAndSpecialCharactersOnly = characters.every(this.isNumberOrSpecialCharacter)

        return isUpperCase && !isNumbersAndSpecialCharactersOnly
    }

    isSilence(value: string): boolean {
        return value === ""
    }

    isNumberOrSpecialCharacter(value: string): boolean {
        return /([\d,. ?])/.test(value)
    }
}

File: scala/bob/src/main/scala/Bob.scala
import Bob.Phrases.{Question, Statement, Yell}

object Bob {
  def response(statement: String): String = statement.trim match {
    case answer @ (Yell(_) | Question(_) | Statement(_)) => answer
    case _ => "Whatever."
  }

  object Phrases {
    object Statement {
      def unapply(arg: String): Option[String] = if (arg.isEmpty) Some("Fine. Be that way!") else None
    }

    object Yell {
      def unapply(arg: String): Option[String] = if (arg.forall(_.isUpper)) Some("Whoa, chill out!") else None
    }

    object Question {
      def unapply(arg: String): Option[String] = if (arg.endsWith("?")) Some("Sure.") else None
    }

  }
}
